File: analyze.R
#!/usr/bin/env Rscript

library(argparse)
library(futile.logger)
library(jsonlite)
library(R6)

parser <- argparse::ArgumentParser()
parser$add_argument(
    "--pkg"
    , type = "character"
    , help = "Name of the R package to test"
)
parser$add_argument(
   "--output_dir"
   , type = "character"
   , default = getwd()
   , help = "Path to write files to"
)

# Grab args
args <- parser$parse_args()

# lil helper
.log_info <- function(msg){
    futile.logger::flog.info(msg)
    return(invisible(NULL))
}

# Store in constants for easier debugging
PKG_NAME <- args[["pkg"]]
OUT_DIR <- args[["output_dir"]]

# grab just the exported objects
.log_info(sprintf("Loading up namespace for package %s", PKG_NAME))
pkg_env <- loadNamespace(PKG_NAME)
export_names <- names(pkg_env[[".__NAMESPACE__."]][["exports"]])
.log_info(sprintf("Found %i exported objects", length(export_names)))

# Set up skeleton thing
# TODO: handle stuff that isn't just functions or R6 classes
# TODO: figure out how to handle S3Methods
# TODO: figure out and document how to handle Python magic methods
out <- list(
    "language" = 'r'
    , "functions" = list()
    , "classes" = list()
)

for (obj_name in export_names){
    obj <- get(obj_name, envir = pkg_env)

    if (is.function(obj)){
        out[["functions"]][[obj_name]] <- list(
            "args" = names(formals(obj))
        )
        next
    }

    if (R6::is.R6Class(obj)){
        out[["classes"]][[obj_name]] <- list()
        next
    }
}

# write it out
out_file <- file.path(OUT_DIR, sprintf("%s.json", PKG_NAME))
.log_info(sprintf("Writing output to %s", out_file))
write(x = jsonlite::toJSON(out), file = out_file, append = FALSE)
.log_info("Done analyzing this package.")

File: cli/doppel/bin/analyze.py
#!/usr/bin/env python3

import argparse
import inspect
import json
import os
import types

parser = argparse.ArgumentParser()
parser.add_argument(
    "--pkg"
    , type=str
    , help="Name of the python package to test"
)
parser.add_argument(
   "--output_dir"
   , type=str
   , default=os.getcwd()
   , help="Path to write files to"
)

# Grab args (store in constants for easier debugging)
args = parser.parse_args()
PKG_NAME = args.pkg
OUT_DIR = args.output_dir
LANGUAGE = 'python'

# Import that module
top_level_env = __import__(PKG_NAME)

# Set up the thing
out = {
    "name": "{} [python]".format(PKG_NAME),
    "language": "python",
    "functions": {},
    "classes": {}
}

# lil helper
def _log_info(msg):
    print(msg)

modules_to_parse = [top_level_env]

while len(modules_to_parse) > 0:

    # Grab the next module
    pkg_env = modules_to_parse.pop()

    # Get the exported stuff
    export_names = list(filter(lambda x: not x.startswith('_'), dir(pkg_env)))

    for obj_name in export_names:
        # Grab the object
        obj = getattr(pkg_env, obj_name)
        # Is it a function?
        if isinstance(obj, types.FunctionType):
            out["functions"][obj_name] = []
            next
        # Is it a class?
        elif inspect.isclass(obj):
            # Is it an exception? (skip)
            if issubclass(obj, Exception):
                _log_info("{} is an Exception. Skipping.".format(obj_name))
            else:
                out['classes'][obj_name] = []
            next
        elif isinstance(obj, types.ModuleType):
            _log_info("{} is a module".format(obj_name))
            modules_to_parse.append(obj)
        else:
            _log_info("Could not figure out what {} is".format(obj_name))

# write it out
out_file = os.path.join(OUT_DIR, "{}_{}.json".format(LANGUAGE, PKG_NAME))
_log_info("Writing output to {}".format(PKG_NAME))
with open(out_file, 'w') as f:
    f.write(json.dumps(out))
_log_info("Done analyzing this package.")
